sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala

```diff
@@ -4021,6 +4021,27 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
       }
     }
   }
+
+  test("SPARK-33474: Support TypeConstructed partition spec value") {
+    withTable("t1", "t2", "t4") {
+      sql("CREATE TABLE t1(name STRING, part DATE) USING PARQUET PARTITIONED BY (part)")
+      sql("INSERT INTO t1 PARTITION(part = date'2019-01-02') VALUES('a')")
+      checkAnswer(sql("SELECT name, CAST(part AS STRING) FROM t1"), Row("a", "2019-01-02"))
+
+      sql("CREATE TABLE t2(name STRING, part TIMESTAMP) USING PARQUET PARTITIONED BY (part)")
+      sql("INSERT INTO t2 PARTITION(part = timestamp'2019-01-02 11:11:11') VALUES('a')")
+      checkAnswer(sql("SELECT name, CAST(part AS STRING) FROM t2"), Row("a", "2019-01-02 11:11:11"))
+
+      val e = intercept[AnalysisException] {
+        sql("CREATE TABLE t3(name STRING, part INTERVAL) USING PARQUET PARTITIONED BY (part)")
+      }.getMessage
+      assert(e.contains("Cannot use interval for partition column"))
+
+      sql("CREATE TABLE t4(name STRING, part BINARY) USING CSV PARTITIONED BY (part)")
+      sql(s"INSERT INTO t4 PARTITION(part = X'537061726B2053514C') VALUES('a')")
+      checkAnswer(sql("SELECT name, cast(part as string) FROM t4"), Row("a", "Spark SQL"))
+    }
```

Comment:
The last thing that I'm concerned about is whether we already have tests corresponding to the @cloud-fan 's last comment.

https://github.com/apache/spark/pull/30421#issuecomment-734049999
```
Let's make sure this feature works correctly:

 - All the literals are supported. Non-literals are forbidden. e.g. part_col=array(1) does not create a string value "array(1)".
 - Null literal is supported. We should use null instead of "null" to represent it.
 - If the literal data type doesn't match the partition column data type, we should do type check and cast like normal table insertion.
```

Reply:
For  these three concern
>  All the literals are supported. Non-literals are forbidden. e.g. part_col=array(1) does not create a string value "array(1)".

Now type constructor only support `DATE` `TIMESTAMP`  `INTERVAL` `X`, so that's not a concern.

 >  Null literal is supported. We should use null instead of "null" to represent it.

It's solved  since #30538 

 >  If the literal data type doesn't match the partition column data type, we should do type check and cast like normal table insertion.

Seems there is auto type conversion. I will check this with UT

Reply:
Hmm Add some UT case. Got the concern that since now we treat partition spec use string, it's not  type safe.
I have chat with @cloud-fan  that we can make partition sepc's value as literal. Then it will be more type safe.

So should we still continue this pr or just start work on treat partition value as Literal?